contracts/v2/EIP3009.sol

```diff
@@ -29,36 +29,28 @@ import { EIP712Domain } from "./EIP712Domain.sol";
 import { EIP712 } from "../util/EIP712.sol";
 
 /**
- * @title Gas Abstraction
- * @notice Provide internal implementation for gas-abstracted transfers and
- * approvals
+ * @title EIP-3009
+ * @notice Provide internal implementation for gas-abstracted transfers
  * @dev Contracts that inherit from this must wrap these with publicly
  * accessible functions, optionally adding modifiers where necessary
  */
-abstract contract GasAbstraction is AbstractFiatTokenV2, EIP712Domain {
+abstract contract EIP3009 is AbstractFiatTokenV2, EIP712Domain {
+    // keccak256("TransferWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)")
     bytes32
         public constant TRANSFER_WITH_AUTHORIZATION_TYPEHASH = 0x7c7c6cdb67a18743f49ec6fa9b35f50d52ed05cbed4cc592e13b44501c1a2267;
-    // = keccak256("TransferWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)")
-    bytes32
-        public constant APPROVE_WITH_AUTHORIZATION_TYPEHASH = 0x808c10407a796f3ef2c7ea38c0638ea9d2b8a1c63e3ca9e1f56ce84ae59df73c;
-    // = keccak256("ApproveWithAuthorization(address owner,address spender,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)")
-    bytes32
-        public constant INCREASE_ALLOWANCE_WITH_AUTHORIZATION_TYPEHASH = 0x424222bb050a1f7f14017232a5671f2680a2d3420f504bd565cf03035c53198a;
-    // = keccak256("IncreaseAllowanceWithAuthorization(address owner,address spender,uint256 increment,uint256 validAfter,uint256 validBefore,bytes32 nonce)")
+
+    // keccak256("ReceiveWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)")
     bytes32
-        public constant DECREASE_ALLOWANCE_WITH_AUTHORIZATION_TYPEHASH = 0xb70559e94cbda91958ebec07f9b65b3b490097c8d25c8dacd71105df1015b6d8;
-    // = keccak256("DecreaseAllowanceWithAuthorization(address owner,address spender,uint256 decrement,uint256 validAfter,uint256 validBefore,bytes32 nonce)")
+        public constant RECEIVE_WITH_AUTHORIZATION_TYPEHASH = 0xd099cc98ef71107a616c4f0f941f04c322d8e254fe26b3c6668db87aae413de8;
+
+    // keccak256("CancelAuthorization(address authorizer,bytes32 nonce)")
     bytes32
         public constant CANCEL_AUTHORIZATION_TYPEHASH = 0x158b0a9edf7a828aad02f63cd515c68ef2f50ba807396f6d12842833a1597429;
-    // = keccak256("CancelAuthorization(address authorizer,bytes32 nonce)")
-
-    enum AuthorizationState { Unused, Used, Canceled }
 
     /**
-     * @dev authorizer address => nonce => authorization state
+     * @dev authorizer address => nonce => bool (true if nonce is used)
      */
-    mapping(address => mapping(bytes32 => AuthorizationState))
-        private _authorizationStates;
```

Comment:
I've verified that this change is OK, EVM treats `0` as `false` and non-zero values as `true`.

Reply:
I tested it manually. I also tested via calling `SLOAD` directly:

```solidity
pragma solidity 0.6.12;

contract Foo {
    enum AuthorizationState { Unused, Used, Canceled }
    
    AuthorizationState public a = AuthorizationState.Canceled;
    
    function authorizationState() external view returns (AuthorizationState) {
        return a;
    }
    
    function authorizationStateBool() external view returns (bool) {
        bool x;
        
        assembly {
            x := sload(0)
        }
        return x;
    }
}
```